### DISCUSSION TITLE
I don't understand a test failure

### DISCUSSION THREAD

**Discussion by easiwriter** [pointfreeco/swift-composable-architecture#3392] [Q&A]
I am struggling to get a test to work. It is modelled after a similar test in SyncUps. The test that fails is testAddProjectButtonTapped at line 22. I have provided the relevant sources in the archive. It also contains the failed state change in Results.rtf.  

What I thought happens is that the project is assigned to the ProjectForm.project and that this should work. What appears to happen is the test doesn't like the project for some reason. My suspicion is that it has something to do with SwiftData because everything works if I take SwiftData out of the equation and use structs without the relationships. Maybe it is the way SwiftData stores model data is upsetting the test framework?

The app itself functions as expected.

[Archive.zip](https://github.com/user-attachments/files/17025183/Archive.zip)

**Comment by mbrandonw** [comment]
Hi @easiwriter, you should provide the full compiling project in order for someone to help. Providing only a few files makes it difficult for someone to understand how everything connects together. And if you can't provide the project, then please reproduce the problem in a minimal project and share that.

**Reply by easiwriter** [reply]
Hi Brandon, thanks for the reply. I attach two zip files. The one called fails.zip contains the failing test. The other called passes.zip works because it uses structs for the models and not classes. I hope you can make sense of these. The project is in an early state - I am converting an existing unfinished SwiftData  project containing lots of code. Why? Because I'm very impressed with TCA, especially the testing features which I've never been good at. No pain, no gain as they say.

[Fails.zip](https://github.com/user-attachments/files/17028601/Fails.zip)
[passes.zip](https://github.com/user-attachments/files/17028605/passes.zip)

**Reply by mbrandonw** [reply]
Hi @easiwriter, this assertion is failing:

```swift
await store.send(.addButtonTapped) {
  $0.addProject?.project = project
  $0.addProject?.projectModel = project.fromProject()
}
```

…because `addProject` is `nil`, and so assigning through `addProject?.` is just a no-op. The test failure message tries to explain this:

```
🛑 Issue recorded: A state change does not match expectation: …

      ProjectsList.State(
        _context: ModelContext(…),
    −   _addProject: nil,
    +   _addProject: ProjectForm.State(
    +     …
    +   )

(Expected: −, Actual: +)
```

This message is saying that you asserted that `addProject` was `nil`, but in reality it was non-`nil`. Is there something that could be improved with this messaging?

To fix need to assign all of `addProject` in this assertion:

```swift
await store.send(.addButtonTapped) {
  $0.addProject = …
}
```

And then this assertion is failing:

```swift
await store.send(.saveProjectButtonTapped) {
  $0.projects = [project]
  $0.addProject = nil
}
```

…because `projectModels` changed but you didn't assert on those state changes. Again the test failure message tries to explain this:

```
🛑 Issue recorded: A state change does not match expectation: …

      ProjectsList.State(
        _context: ModelContext(…),
        _addProject: nil,
        _projectModels: #1 IdentifiedArray(
    +     ProjectModel(
    +       _id: ProjectModel._SwiftDataNoType(),
    +       _name: ProjectModel._SwiftDataNoType(),
    +       _type: ProjectModel._SwiftDataNoType(),
    +       _listMarkers: ProjectModel._SwiftDataNoType(),
    +       _paperName: ProjectModel._SwiftDataNoType(),
    +       _timeStamp: nil,
    +       _pageSetup: nil,
    +       _styleSheet: nil,
    +       _folders: nil,
    +       _textFiles: nil
    +     )
        ),
        _projects: #2 […],
        _templates: #3 […]
      )

(Expected: −, Actual: +)
```

I will also say that unfortunately Swift Data does not immediately play nicely with TCA's testing tools. Our tools are mostly built on the properties of value types because they can be copied before and after an action is sent so that we can compare what changed. And unfortunately Swift Data is mostly built on reference types, which are much more difficult (or often _impossible_) to unit test. So in general you are going to have a very difficult time testing Swift Data in a TCA app.

**Comment by easiwriter** [comment]
Hi Brandon

Thanks for this, which confirms my suspicion about SwiftData. I will investigate ways out of the impasse. When do you think TCA will handle this, if ever?

**Reply by easiwriter** [reply]
Hi Brandon, the SwiftData app I mentioned in my original post was built using a ViewModel technique that avoids having SwiftData code in a View, which can and usually does cause problems with state changes. I have now ported the ViewModel code to the new app and removed the class references from the Reducer. By refactoring this way the Reducers have no direct access to the database classes: instead the Reducer updates the Database via the ViewModel. The test now works OK. Here is a copy of the project should you feel inclined to look at it.

[Writing Shed 2.zip](https://github.com/user-attachments/files/17042105/Writing.Shed.2.zip)
